feat: Add Coordinator Waffle Flags

grmartin opened the pull request:
- fix: swapping out getQueryParameters with searchParams

  - `getQueryParameters` has been deprecated, so the code has been modified to use `URL(...).searchParams`.

  - Moving Waffle Flag handling to an external file.

  - 100% Coverage for Waffle Flag processing as well as interception.

    Also allowlisted a vulnerability.

    > Note: Used API 2 for Jest in its Tabular format. Nice for simpler values. and less eslint/visual formatting issues.

- fix: Merge COMMERCE_COORDINATOR_BASE_URL in Config from Env

  - Includes moving service code so CC calls are at the end of the file.

- chore: Fixing some excluded code coverage files

  - Permitting `src/data` to be covered
    - excluding the specific files without coverage.
  - Added a simple rootSagas Test

- fix: Moving isWaffleFlagEnabled with its sibling function

- feat: Accepting Waffle Flags from the Environment 

  - A simple hijack of URL parsing to allow flags from the environment. This should ease development.

- feat: Adding isCommerceCoordinatorEnabled

  - Checks Waffle Flags to see if were in CC mode.

    Additional testing and test refactoring.

- chore: Doc updates and moving Waffle Flags from magic strings to an Enum

- chore: Set CC Waffle Flag to use the MS1 Switch for Backend Selection

- feat: Swapping between Ecomm IDA and Coordinator based on Flags

  - Testing updates including returning promises from the `perform` functions so that async work doesn't time out Jest
  - Documentation Updates
  - Setting payment processing dialog to never show unless we are in CC Mode
  
  JIRA: https://2u-internal.atlassian.net/browse/THES-49

grmartin committed:
fix: swapping out `getQueryParameters` with `searchParams`

- `getQueryParameters` has been deprecated, so the code has been modified to use `URL(...).searchParams`.
- Moving Waffle Flag handling to an external file.
- 100% Coverage for Waffle Flag processing as well as interception.

Also allowlisted a vulnerability.

> Note: Used API 2 for Jest in its Tabular format. Nice for simpler values. and less eslint/visual formatting issues.

grmartin committed:
fix: Merge `COMMERCE_COORDINATOR_BASE_URL` in Config from Env

- Includes moving service code so CC calls are at the end of the file.

grmartin committed:
chore: Fixing some excluded code coverage files

- Permitting `src/data` to be covered
    - excluding the specific files without coverage.
- Added a simple rootSagas Test

grmartin committed:
fix: Moving `isWaffleFlagEnabled` with its sibling functions

- Also, testing.

grmartin committed:
feat: Accepting Waffle Flags from the Environment

- A simple hijack of URL parsing to allow flags from the environment. This should ease development.

grmartin committed:
feat: Adding `isCommerceCoordinatorEnabled`

Checks Waffle Flags to see if were in CC mode.

Additional testing and test refactoring.

grmartin committed:
chore: Doc updates and moving Waffle Flags from magic strings to an Enum

grmartin committed:
chore: Set CC Waffle Flag to use the MS1 Switch for Backend Selection

grmartin committed:
feat: Swapping between Ecomm IDA and Coordinator based on Flags

- Testing updates including returning promises from the `perform` functions so that async work doesn't time out Jest
- Documentation Updates
- Setting payment processing dialog to never show unless we are in CC Mode

grmartin commented on src/payment/data/sagas.test.js:
```diff
@@ -41,14 +41,15 @@ import '../__factories__/basket.factory';
   DEFAULT_PAYMENT_STATE_POLLING_MAX_ERRORS,
   PAYMENT_STATE,
-  POLLING_PAYMENT_STATES,
+  POLLING_PAYMENT_STATES, WAFFLE_FLAGS,
```
fix for this formatting issue is already complete and will be pushed with other preliminary feedback.

grmartin committed:
fix: Remove CyberSource call from Coordinator Service

pshiu commented on src/payment/data/utils.js:
```diff
@@ -288,3 +284,14 @@ export function createCustomRoutine(name, addtlStages, keepDefaults = true) {
     ...addtlStages.map(x => x.toUpperCase()),
   ])(name);
 }
+
+/**
+ * Determines if the MFE should follow the Commerce Coordinator Flows
+ *
+ * This checks a waffle flag and assumes the CommerceCoordinator has set the flag if the order type is right.
+ *
+ * @returns {boolean}
+ */
+export function isCommerceCoordinatorEnabled() {
+  return isWaffleFlagEnabled(WAFFLE_FLAGS.COMMERCE_COORDINATOR_ENABLED);
+}
```
Can we call `isWaffleFlagEnabled(WAFFLE_FLAGS.COMMERCE_COORDINATOR_ENABLED)` directly? Worried this is too much wrapping.

pshiu commented on src/data/waffleFlags.js:
```diff
@@ -0,0 +1,55 @@
+import { getConfig } from '@edx/frontend-platform';
```
Can we add a link to [Feature Toggle HOWTO](https://github.com/openedx/frontend-app-payment/blob/master/docs/how_tos/feature_toggle.rst) somewhere appropriate in this file?

pshiu commented on src/payment/data/constants.js:
```diff
@@ -1,11 +1,48 @@
+/**
+ * Order Types
```
```suggestion
 * Order Types are the possible types of products a learner can purchase.
 *
 * Each product type may have different user flow requirements.
 *
 * For example, bulk enrollments allow users to change the quantity of a product.
```

pshiu commented on src/payment/data/constants.js:
```diff
@@ -1,11 +1,48 @@
+  /**
+   * The Redemption of a Single Course Enrollment Code
+   *
+   * A bulk-purchases on a single course, usually B2C for 100% off of course seats.
```
Spelling:

```suggestion
   * A bulk purchase on a single course, usually B2C for 100% off of course seats.
```

pshiu commented on src/payment/data/constants.js:
```diff
@@ -1,11 +1,48 @@
+  /**
+   * Program Entitlement
+   *
+   * This means someone has purchased a full program containing the course, so they have already paid but may not
+   * actually be enrolled in a run.
```
```suggestion
   * A purchase of a program, which are multiple products called "course entitlements".
   * 
   * A course entitlement allows a learner to redeem the entitlement for a seat in a course at a future time.
```

pshiu commented on src/payment/data/constants.js:
```diff
@@ -15,6 +52,9 @@ export const CERTIFICATE_TYPES = {
+ * @readonly
+ * @enum {string|Symbol}
  */
 export const PAYMENT_STATE = (((webserviceEnum = {
```
Nit: Can we replace `// Our Additions` on line 80 below with an explanation that the payment states listed below are added by Payment MFE to the backend's list of possible payment states?

pshiu commented on src/payment/data/service.js:
```diff
@@ -94,18 +93,49 @@ export async function getDiscountData(courseKey) {
+// Commerce Coordinator Specific Calls
+const CC_BASE = getConfig().COMMERCE_COORDINATOR_BASE_URL;
+
+export class CommerceCoordinator {
```
In general, we should try to match Coordinator's API to frontend-app-payment rather than vice-versa. Could you document what we need to do to get Coordinator working with frontend-app-payment?

If there is a refactor we can do to reduce the amount of multiplexing here, that would be awesome.

grmartin committed:
chore: PR Updates, Testing and Removal of `GetActiveOrder`.

- The Service layer of the payment module now shares common data flow.
- GetActiveOrder (THES-234) removed as it has the same flow as getBasket.
- Tests reworked to cover both data flows.

grmartin committed:
fix: Update .env.development-stage to target devstack

- Target DevStack CC unless overridden by `.env.private`

Co-authored-by: Phillip Shiu <pshiu@edx.org>

grmartin committed:
chore: PR Updates

Update src/data/waffleFlags.js

Co-authored-by: Phillip Shiu <pshiu@edx.org>

Update .env.test

Co-authored-by: Phillip Shiu <pshiu@edx.org>

Update src/data/waffleFlags.js

Co-authored-by: Phillip Shiu <pshiu@edx.org>

pshiu commented on src/data/waffleFlags.test.js:
```diff
@@ -0,0 +1,167 @@
+/**
+ * Set our JSDOM Window and Document Location
+ *
+ * **NOTE**: You *MUST* include your tests async DoneCallback as the final action of your `perform`'s closure.
+ * failure to do so will lead to test timeouts.
```
Question: Why is this comment located here, as opposed to the test files where `done()` is called?

Asking because placing this comment here makes it sound like there is something in this function that requires a done callback, but I'm not sure what it is.

My understanding was that the done callback pattern is only necessary when we have a test with a callback we don't await. But what callback are we are not awaiting in our tests - could you point me to them? Or am I missing something?

grmartin replied:
I have added to the documentation, feel free to open this again if its confusing.

pshiu commented on .env:
```diff
@@ -38,4 +38,5 @@ STRIPE_RESPONSE_URL=null
-COMMERCE_COORDINATOR_BASE_URL=null
\ No newline at end of file
+COMMERCE_COORDINATOR_BASE_URL=null
+WAFFLE_FLAGS=""
```
We should note somewhere that waffle flags coming via query parameters override any `WAFFLE_FLAGS` set here.

grmartin replied:
agreed, no docs have been updated. with this work. There is a CC side ticket for this, ill mod it to be both.

grmartin replied:
[THES-242](https://2u-internal.atlassian.net/browse/THES-242)

pshiu commented on src/payment/data/selectors.js:
```diff
@@ -90,9 +91,11 @@ export const paymentProcessStatusSelector = state => (state[storeName].basket.pa
  */
 export const paymentProcessStatusShouldRunSelector = state => (
   POLLING_PAYMENT_STATES.includes(paymentProcessStatusSelector(state))
+  && isWaffleFlagEnabled(WAFFLE_FLAGS.COMMERCE_COORDINATOR_ENABLED)
```
The collision of the constant `WAFFLE_FLAGS` with the environmental variable `WAFFLE_FLAGS` set in .env & sister files got confusing for me. What do you think of renaming one of them?

grmartin replied:
Environment is simpler, ill do it there... `ENABLED_WAFFLE_FLAGS`?

pshiu commented on src/payment/data/service.js:
```diff
@@ -3,12 +3,68 @@ import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
   'LMS_BASE_URL',
 ], 'payment API service');
 
+/* eslint-disable no-use-before-define */ /* Chicken vs Egg...
+    but it is preferred the interpreter pickup errors rather than use magic strings */
```
Might be worth elaborating what is the chicken vs. egg problem with `.name`.

grmartin replied:
I assumed general knowlege of our tools, ill simplify.

pshiu commented on src/payment/data/service.js:
```diff
@@ -3,12 +3,68 @@ import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
+ *
+ * @type {{[p: string]: string}}
+ */
+const urlsByFunction = {
```
It sounds like this will require filing tickets for remapping the Coordinator & API specs to use similar endpoints names as Ecommerce?

What if we want to name something differently in the future?

grmartin replied:
We did that in PR 100. The documentation is wrong now. Was your suggestion.

pshiu commented on src/payment/data/service.js:
```diff
@@ -3,12 +3,68 @@ import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
+ */
+const urlsByFunction = {
+  /* Shared common paths */
+  [getClientSecret.name]: '/bff/payment/v0/capture-context', // no forward slash at the end or we will explode
```
"We will explode" is colorful, but not really accurate. Could we say that ecommerce will 404?

grmartin replied:
i dont recall the exact error, but im coonfusing whether its BAD_METHOD or FILE_NOT_FOUND. But i will clarify.

pshiu commented:
@grmartin, could you add a commitlint prefix to the PR title & add a link to the ticket in the description of the PR?

grmartin commented on .env:
```diff
@@ -38,4 +38,5 @@ STRIPE_RESPONSE_URL=null
-COMMERCE_COORDINATOR_BASE_URL=null
\ No newline at end of file
+COMMERCE_COORDINATOR_BASE_URL=null
+WAFFLE_FLAGS=""
```
habit. ill fix

grmartin committed:
fix: PR Update, Update src/data/waffleFlags.js

Co-authored-by: Phillip Shiu <pshiu@edx.org>

grmartin commented on src/data/waffleFlags.test.js:
```diff
@@ -0,0 +1,167 @@
+/**
+ * Set our JSDOM Window and Document Location
+ *
+ * **NOTE**: You *MUST* include your tests async DoneCallback as the final action of your `perform`'s closure.
+ * failure to do so will lead to test timeouts.
```
Fair, but that is arguably a flaw in our package.json file and we all should be when able (makes jumping in and fixing things easier) But happy to fix.

grmartin commented on src/payment/data/reducers.js:
```diff
@@ -141,6 +135,7 @@ const clientSecretInitialState = {
   clientSecretId: '',
 };
 
+// CC & Ecomm IDA Backends, Stripe
```
fair.

grmartin commented on src/payment/data/sagas.js:
```diff
@@ -41,8 +44,13 @@ import { checkout as checkoutPaypal } from '../payment-methods/paypal';
+  POLLING_PAYMENT_STATES,
+} from './constants';
 import { generateApiError } from './handleRequestError';
+import { getCurrentPaymentState } from './service';
```
Yes, my IDE is configured to not use *s id assume, since its considered bad form..

grmartin commented on src/data/waffleFlags.test.js:
```diff
@@ -0,0 +1,167 @@
+    //   This was WAY TOO HARD FOUGHT.
+    const testLocation = `${window.origin}/dox.asp?${WAFFLE_PREFIX}xyzzy=on`;
+
+    return performWithModifiedJSDOMLocation(testLocation, () => {
+      const result = processUrlWaffleFlags();
+      expect(result).toStrictEqual({ xyzzy: true });
+    });
```
No, the machoism for handling history in Jest doesnt work right as it mimics a full DOM via JSDOM. Wes and i tried some 7 different forms, none took. This is why we had to work with the history scoping mechanism as well because it throws various `SecurityError: `'s, such as the one depper in the stack called `replaceState` even though we never touch it directly, its minicing the Mozilla Docs (i assume) and such.

grmartin replied:
Also theres needed cleanup, to reset the DOM for the next call, this requires the caller to know nothing of that, and it 'just works'.

grmartin commented on src/data/waffleFlags.test.js:
```diff
@@ -0,0 +1,167 @@
+
+describe('getWaffleFlags', () => {
+  it('should default to document.location when empty', () => {
+    // We have to use the Window's origin, otherwise: SecurityError: replaceState cannot update...
```
In code update, eopen if still an issue :)

grmartin committed:
chore: PR Updates

grmartin committed:
chore: typo

grmartin committed:
chore: Added null check to WaffleFlags
